src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -33,47 +38,37 @@ internal sealed partial class JsonClassInfo
         // If enumerable, the JsonClassInfo for the element type.
         internal JsonClassInfo ElementClassInfo { get; private set; }
 
+        public JsonSerializerOptions Options { get; private set; }
+
         public Type Type { get; private set; }
 
         internal void UpdateSortedPropertyCache(ref ReadStackFrame frame)
         {
-            // Todo: on classes with many properties (about 50) we need to switch to a hashtable for performance.
-            // Todo: when using PropertyNameCaseInsensitive we also need to use the hashtable with case-insensitive
-            // comparison to handle Turkish etc. cultures properly.
-
-            Debug.Assert(_propertyRefs != null);
-
-            // Set the sorted property cache. Overwrite any existing cache which can occur in multi-threaded cases.
-            if (frame.PropertyRefCache != null)
+            // Check if we are trying to build the sorted cache.
+            if (frame.PropertyRefCache == null)
             {
-                HashSet<PropertyRef> cache = frame.PropertyRefCache;
-
-                // Add any missing properties. This creates a consistent cache count which is important for
-                // the loop in GetProperty() when there are multiple threads in a race conditions each generating
-                // a cache for a given POCO type but with different property counts in the json.
-                while (cache.Count < _propertyRefs.Count)
-                {
-                    for (int iProperty = 0; iProperty < _propertyRefs.Count; iProperty++)
-                    {
-                        PropertyRef propertyRef = _propertyRefs[iProperty];
-                        // Cache the missing property or override the existing property.
-                        cache.Add(propertyRef);
-                    }
-                }
+                return;
+            }
 
-                Debug.Assert(cache.Count == _propertyRefs.Count);
-                if (_propertyRefsSorted == null || _propertyRefsSorted.Length < cache.Count)
-                {
-                    _propertyRefsSorted = new PropertyRef[cache.Count];
-                }
-                cache.CopyTo(_propertyRefsSorted);
-                frame.PropertyRefCache = null;
+            List<PropertyRef> newList;
+            if (_propertyRefsSorted != null)
+            {
+                newList = new List<PropertyRef>(_propertyRefsSorted);
```

Comment:
Why create a List? It would be more efficient to create the array directly and copy the values over.

Reply:
It was a guard for multi-threading. This way the `_propertyRefsSorted` is a single operation

Reply:
> This way the _propertyRefsSorted is a single operation

I don't understand, the List<T> constructor is still doing multiple steps...

Reply:
The instance of the array is read\written by multiple threads and assigned to in a single step. If we instead copy directly to the array that could have overlapping writes. The List is not shared by multiple threads.